JAVA-2967: Support native_transport_(address|port) + native_transport_port_ssl for DSE 6.8

absurdfarce opened the pull request.

Commit:
Basic fix, seems to make local repro case happy.  Still need to get to work on some tests.

absurdfarce commented:
As of this writing this is still an in-progress PR.  The basic fix contained within seems to resolve the issue in my local repro case but no tests have been added yet for the new functionality.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -956,7 +975,11 @@ private boolean isValidPeer(Row peerRow, boolean logIfInvalid) {
+              || (peerRow.getColumnDefinitions().contains("native_transport_address")
+                  && peerRow.getColumnDefinitions().contains("native_transport_port")
+                  && !peerRow.isNull("native_transport_address")
+                  && !peerRow.isNull("native_transport_port"));
```
How about a row that has `native_transport_port_ssl` but not `native_transport_port`? Can this even happen?

absurdfarce replied:
native_transport_port has a default value of 9042 so even if nothing is specified for it in cassandra.yaml the corresponding system.peers row will still have a value.  I don't think there is a way to create a row here that doesn't have a native_transport_port entry so this logic should hold up.  I'll take a look at DSE code in a bit to confirm.

I also just fired up my local 3 node DSE 6.8.13 cluster (with no value specified for native_transport_port on any node) to confirm... and it does look like the value is populated as expected.

absurdfarce commented on driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java:
```diff
@@ -611,6 +611,17 @@ private static void updateInfo(
+          && !row.isNull("native_transport_port_ssl")) {
+        nativePort = row.getInt("native_transport_port_ssl");
+      }
+      broadcastRpcAddress = new InetSocketAddress(nativeAddress, nativePort);
```
Location of this block (and similar logic in the other blocks) was very deliberate.  We can't go underneath the base case handler of rpc_address + default port but it does seem like this specific case will be less likely than the native_address/port case above us.

Commit:
Minor test fix to account for changes in https://github.com/datastax/java-driver/pull/1571

Commit:
Passing test for C* 4.0, still need to add test coverage for the other two cases

Commit:
Cleaning up the contains test a bit

Commit:
Base tests should be covered, still need to fix rpc_address case which should use Scassandra ports directly

Commit:
Base tests for all three cases are passing now.  Still need an additional test to validate support for
native_transport_port_ssl when it's set.

Commit:
Some minor fixes

absurdfarce commented on driver-core/src/test/java/com/datastax/driver/core/exceptions/ExceptionsTest.java:
```diff
@@ -284,7 +284,7 @@ public void should_create_proper_read_timeout_exception() {
     assertThat(e.getMessage())
         .isEqualTo(
-            "Cassandra timeout during read query at consistency LOCAL_QUORUM (3 responses were required but only 2 replica responded)");
+            "Cassandra timeout during read query at consistency LOCAL_QUORUM (3 responses were required but only 2 replica responded). In case this was generated during read repair, the consistency level is not representative of the actual consistency.");
```
This and the similar change below address some minor fallout from https://github.com/datastax/java-driver/pull/1571

adutra replied:
Ouch, and I approved that PR :-D.
Alternatively, you could use `assertThat(...).contains(...)` to keep the code concise.

absurdfarce replied:
Hmmmm... that might actually be desirable anyway.  It makes the test a purer expression of the thing we're actually testing for rather than worrying about the total structure of the log message.  I think I like that.

absurdfarce commented on driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java:
```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+    }
+  }
+
+  static class PeerRowState {
```
State object + builder is a bit more convoluted than I'd like but it has the effect of nicely isolating changes required for an individual test.  Without this class I found myself passing around multiple config options (all of which were either related or similar to each other) in order to setup a given test.  Introducing this class had the effect of making the intent of a single test case above very clear.

adutra replied:
Very clean design imo :-)

absurdfarce commented on driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java:
```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+   */
+  @Test(groups = "short")
+  @CCMConfig(createCcm = false)
+  @Ignore("Requires SSL support in scassandra")
```
Really no good way to test this without SSL support for scassandra.  https://github.com/scassandra/scassandra-server/issues/65 is of relevance here.

A minor addendum; this might be obvious but any change here would also require tweaks to runPeerTest() in order to allow it to (optionally) enable SSL based on the contents of the PeerRowState instance.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java:
```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+    PeerRowState state =
+        PeerRowState.builder()
+            .peers("rpc_address", expectedAddress)
+            /* DefaultEndPointFactory isn't happy if we don't have both a value for
```
```suggestion
            /* DefaultEndPointFactory isn't happy if we don't have a value for
```

adutra commented on driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java:
```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+
+  /**
+   * If both native_transport_port and native_transport_port_ssl are present we expect the SSL
+   * version to win. Note that we don't make this conditional on SSL actually being configured.
```
> Note that we don't make this conditional on SSL actually being configured

Are we sure? I see SSL being tested in a few places with the expression:

```
cluster.getCluster().getConfiguration().getProtocolOptions().getSSLOptions() != null
```

absurdfarce replied:
Agreed.  I refreshed my recollection after writing that comment and just forgot to go clean it up.  I'll clean this up so that it makes sense.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java:
```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+      private InetAddress expectedAddress;
+      private Optional<Integer> expectedPort = Optional.absent();
+
+      private BitSet shouldPrime = new BitSet(3);
```
What is the third bit used for? If only 2 are required, maybe you could simplify and use 2 booleans?

absurdfarce replied:
This is also a holdover that prolly doesn't make sense anymore.  The third bit originally referred to system.local, but as discussed above we have to _always_ prime system.local now.

It also probably does make sense to just switch to booleans now.  With three vals it seemed worth encapsulating everything in a single data structure... but by the time you add in static flags for each name the benefit is somewhat lost.

adutra reviewed (changes requested):
Looking really good!

absurdfarce committed:
Update driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java

Co-authored-by: Alexandre Dutra <adutra@users.noreply.github.com>

Commit:
Feedback from code review

absurdfarce commented:
Thanks for the very complete review @adutra !  I _believe_ I've addressed all the points you raised so this should be ready for another pass.  If I missed something then by all means let me know.

Thanks again!
